### ISSUE TITLE
Solver errors

### ISSUE THREAD

**Issue by erikbrinkman** [quadprog/quadprog#36]
There seems to be a few residual bugs in the solver. I found solutions for them at one point, but unfortunately forgot what I did.

First, a test case that fails is [here](https://github.com/albertosantini/node-quadprog/blob/master/test/barbiggs-data.json).

Second, in `verify` you only check that inequality constraints have a non-negative lagrange multiplier, but I believe all the multipliers should be non-negative. If you take a test case with several equality constraints, you should be able to flip the signs on any constraint and get the same multiplier, but you don't with this implementation, naturally I think this has to do with how reverse step is handled, maybe here https://github.com/quadprog/quadprog/blob/a404dfbdb4ab635e7930f41e1b2a6418b6ba4709/quadprog/solve.QP.c#L239-L241. My general feeling is that this is also related to the problem with "barbiggs".

Lastly, this isn't an issue with accuracy, but the check for `reverse_step` here https://github.com/quadprog/quadprog/blob/a404dfbdb4ab635e7930f41e1b2a6418b6ba4709/quadprog/solve.QP.c#L259 is unnecessary as it will never be set for inequality constraints.

Sorry this issue isn't more helpful.

**Comment by preeyan** [comment]
Thanks for raising these points @erikbrinkman.

In regard to the first point:

I used quadprog v0.1.11 to solve the problem at https://github.com/albertosantini/node-quadprog/blob/master/test/barbiggs-data.json. The correct solution appears to be achieved. The verify method in its current form fails, but only because the precision is too tight for this test case. It passes if the the 1e-15 is adjusted to 1e-10 on this line: https://github.com/quadprog/quadprog/blob/a404dfbdb4ab635e7930f41e1b2a6418b6ba4709/tests/test_1.py#L47

In regard to the second point:

Ultimately the choice of sign of the Lagrange multipliers is a matter of convention. The implementation here matches the behaviour of MATLAB and Octave reported in https://github.com/quadprog/quadprog/issues/2. One property of this convention is that the Lagrange multiplier vector gives the gradient of the optimal objective value with respect to changes in the right-hand sides of the constraints.

In regard to the third point:

I do believe that the full condition here is necessary. Perhaps the easiest way to demonstrate this is to replace it with `if (iact[i] > meq && rv[i] < 0.)` and run the tests. I found that they entered an infinite loop.

In summary: I don't believe that there is anything further to do here, but of course do post back if I have misinterpreted any of your comments.

**Comment by erikbrinkman** [comment]
@preeyan Thanks for the detailed response! It prompted me to go back and analyze all my claims again.

1. You're correct that it's just a tolerance issue. I remember it being more significant, but clearly it's not.
2. Reasonable people can disagree about the sign, but I was also claiming that the magnitude was off. The simple test case for this was:

    ```python
    G = np.eye(5)
    a = np.array([0.73727161, 0.75526241, 0.04741426, -0.11260887, -0.11260887])
    C = np.array([
        [3.6, 0., -9.72],
        [-3.4, -1.9, -8.67],
        [-3.8, -1.7, 0.],
        [1.6, -4., 0.],
        [1.6, -4., 0.]
    ])
    b = np.array([1.02, 0.03, 0.081])
    meq = b.size
    
    prev = None
    for mults in product(*([[1, -1]] * meq)):
        flip = np.array(mults)
        _, _, _, _, lagr, _ = solve_qp(G, a, C * flip, b * flip, meq)
    
        if prev is not None:
            np.testing.assert_array_almost_equal(np.abs(prev), np.abs(lagr))
        prev = lagr
     ```
   but that worked, so again I'm not really sure what I remember, but clearly this is working as intended.
3. the problem with what you tried is that it should be `rv[i] > 0.`. When you try that, all the tests still pass.
